src/NLog/Targets/DatabaseTarget.cs

```diff
@@ -522,6 +524,55 @@ private void WriteEventToDatabase(LogEventInfo logEvent)
             }
         }
 
+        /// <summary>
+        /// Allows the database type to be customized if we
+        /// are not able to properly infer the database type.
+        /// </summary>
+        /// <param name="dbDataParameter">The parameter.</param>
+        /// <param name="parameterInfo"></param>
+        internal void SetParamType(IDbDataParameter dbDataParameter, DatabaseParameterInfo parameterInfo)
+        {
+            if (string.IsNullOrEmpty(parameterInfo.DbType))
+                return;
+            var lastIndexOfDot = parameterInfo.DbType.LastIndexOf('.');
+
+            // Get just the fully qualified name of the enum but not the value
+            var justEnumName = lastIndexOfDot > 0 ? 
+                parameterInfo.DbType.Substring(0, lastIndexOfDot) : string.Empty;
+
+            object rawConstantValue = null;
+            Type paramType = dbDataParameter.GetType();
+            if (!string.IsNullOrEmpty(justEnumName))
+            {
+                // get the enum type from the database parameter we are using
+                var databaseType = paramType.Assembly.GetType(justEnumName);
+
+                // get the enum value for the database type
+                var justEnumValue = parameterInfo.DbType.Substring(lastIndexOfDot + 1,
+                    parameterInfo.DbType.Length - lastIndexOfDot - 1);
+                var typeField = databaseType != null ? databaseType.GetField(justEnumValue) : null;
+                rawConstantValue = typeField != null ? typeField.GetRawConstantValue() : null;
+            }
+
+            // Each custom database type that we want to support will need added here on as needed basis
+            string propertyName = string.Empty;
+            if (parameterInfo.DbType.Contains("OracleDbType"))
+            {
+                propertyName = "OracleDbType";
+            }
+
+            // find the property on the database specific parameter
+            var property = paramType.GetProperty(propertyName);
```

Comment:
I have an exception being thrown for that case with a unit test checking for it to do that. If we fallback to IDataParameter.DbType won't it just cover up a miss configuration instead of helping the user find the issue?